File: src/drift/apply.ts
/**
 * Gate-choice application (R22, R22a, R22b, R23, R23a, R24). Given a resolved
 * GateChoice, dispatch to the right E3 primitive:
 *   - no-drift-proceed  → materialize directly (no backup)
 *   - discard           → snapshotForDiscard + materialize (R23a backup)
 *   - persist           → persistAndMaterialize (R22b transactional pair)
 *   - abort             → no-op (R24)
 *
 * Lock precondition: the caller (E5 swap orchestration) must hold the project
 * lock around this call. We don't acquire it here so that the entire swap
 * sequence — drift detect + gate decide + gate apply + state-write — is
 * bracketed by a single lock acquisition (E3 invariant: lock brackets the
 * rename pair AND the state-write).
 *
 * Read-only callers (the `drift` command, pre-commit warn) don't need this
 * module — they call `detectDrift` directly.
 */

import type { MergedFile } from "../merge/types.js";
import type { ResolvedPlan } from "../resolver/types.js";
import { snapshotForDiscard } from "../state/backup.js";
import { materialize, type MaterializeResult } from "../state/materialize.js";
import type { StatePaths } from "../state/paths.js";
import { persistAndMaterialize } from "../state/persist.js";

import type { GateChoice } from "./types.js";

export interface ApplyGateOptions {
  paths: StatePaths;
  /** New plan (target of the swap). */
  plan: ResolvedPlan;
  /** Merged files for the new plan. */
  merged: ReadonlyArray<MergedFile>;
  /**
   * Active profile name from `.state.json` BEFORE the swap. Required when
   * `choice === "persist"` — that's the profile the live `.claude/` is
   * persisted *into*. Null is acceptable for discard / no-drift / abort
   * paths. Persist with null active is a programmer error (gate decision
   * never returns persist on NoActive).
   */
  activeProfileName: string | null;
}

export type ApplyGateAction =
  | "materialized"
  | "persisted-and-materialized"
  | "aborted";

export interface ApplyGateResult {
  action: ApplyGateAction;
  /** Backup snapshot path if discard branch took one; null otherwise. */
  backupSnapshot: string | null;
  /** Materialize result, or null on abort. */
  materializeResult: MaterializeResult | null;
}

/**
 * Apply the user's (or auto-resolved) gate choice. Returns a structured
 * result so the orchestrator can build user-facing messages without
 * re-deriving state.
 */
export async function applyGate(
  choice: GateChoice,
  opts: ApplyGateOptions,
): Promise<ApplyGateResult> {
  switch (choice) {
    case "abort":
      return { action: "aborted", backupSnapshot: null, materializeResult: null };

    case "no-drift-proceed": {
      const r = await materialize(opts.paths, opts.plan, opts.merged);
      return {
        action: "materialized",
        backupSnapshot: null,
        materializeResult: r,
      };
    }

    case "discard": {
      // Snapshot BEFORE the rename so the backup captures pre-swap content
      // (R23a). materialize() then performs the pending/prior rename — by
      // the time .claude/ is overwritten, the snapshot is already on disk.
      const backup = await snapshotForDiscard(opts.paths);
      const r = await materialize(opts.paths, opts.plan, opts.merged, {}, backup);
      return {
        action: "materialized",
        backupSnapshot: backup,
        materializeResult: r,
      };
    }

    case "persist": {
      if (opts.activeProfileName === null) {
        // Defense-in-depth: decideGate() returns "no-drift-proceed" when
        // fingerprintOk is false (which implies activeProfile is null). If
        // we land here, a caller bypassed the decider with a hand-crafted
        // choice. Refuse rather than write into a profile-less directory.
        throw new Error(
          "persist gate choice requires an active profile in .state.json — none recorded",
        );
      }
      const r = await persistAndMaterialize(opts.paths, {
        activeProfileName: opts.activeProfileName,
        newPlan: opts.plan,
        newMerged: opts.merged,
      });
      return {
        action: "persisted-and-materialized",
        backupSnapshot: null,
        materializeResult: r,
      };
    }
  }
}

File: go/internal/drift/apply.go
package drift

import (
	"errors"

	"github.com/htxryan/c3p/internal/merge"
	"github.com/htxryan/c3p/internal/resolver"
	"github.com/htxryan/c3p/internal/state"
)

// ApplyGateAction is the orchestrator-visible outcome of ApplyGate.
type ApplyGateAction string

const (
	ApplyActionMaterialized            ApplyGateAction = "materialized"
	ApplyActionPersistedAndMaterialized ApplyGateAction = "persisted-and-materialized"
	ApplyActionAborted                 ApplyGateAction = "aborted"
)

// ApplyGateOptions bundles inputs to ApplyGate. ActiveProfileName is the
// active profile name from .state.json BEFORE the swap; required when
// Choice == persist (that's the profile the live .claude/ is persisted
// into). Empty is acceptable for discard / no-drift / abort.
type ApplyGateOptions struct {
	Paths             state.StatePaths
	Plan              resolver.ResolvedPlan
	Merged            []merge.MergedFile
	ActiveProfileName string
}

// ApplyGateResult is the structured outcome so the orchestrator can build
// user-facing messages without re-deriving state.
//
// BackupSnapshot is *string (TS parity with `string | null`): nil signals
// "no snapshot taken" (abort, no-drift, persist, or discard with no live
// .claude/ to back up). Non-nil holds the absolute path of the snapshot dir.
// D7's JSON output emits `null` vs `"<path>"` to surface the distinction.
type ApplyGateResult struct {
	Action            ApplyGateAction
	BackupSnapshot    *string
	MaterializeResult *state.MaterializeResult
}

// ApplyGate dispatches the user's (or auto-resolved) gate choice:
//   - no-drift-proceed → Materialize directly (no backup)
//   - discard          → SnapshotForDiscard + Materialize (R23a backup)
//   - persist          → PersistAndMaterialize (R22b transactional pair)
//   - abort            → no-op (R24)
//
// Lock precondition: caller (D7 swap orchestration) MUST hold the project
// lock around this call. We don't acquire it here so the entire swap
// sequence — drift detect + gate decide + gate apply + state-write — is
// bracketed by a single WithLock. Calling outside a lock is a programmer
// error and breaks the rename-pair atomicity invariant.
//
// PR24 (drift snapshot immutability): the caller MUST NOT re-detect drift
// between display-time and apply-time. The orchestrator runs detect once
// inside the lock and applies the resulting choice; if the user picked from
// a stale outside-lock report, it remains the same DriftReport instance
// passed to display (the gate decision is what the user authorized).
func ApplyGate(choice GateChoice, opts ApplyGateOptions) (ApplyGateResult, error) {
	switch choice {
	case GateChoiceAbort:
		return ApplyGateResult{
			Action:            ApplyActionAborted,
			BackupSnapshot:    nil,
			MaterializeResult: nil,
		}, nil

	case GateChoiceNoDriftProceed:
		r, err := state.Materialize(opts.Paths, opts.Plan, opts.Merged, state.MaterializeOptions{}, nil)
		if err != nil {
			return ApplyGateResult{}, err
		}
		return ApplyGateResult{
			Action:            ApplyActionMaterialized,
			BackupSnapshot:    nil,
			MaterializeResult: &r,
		}, nil

	case GateChoiceDiscard:
		// Snapshot BEFORE the rename so the backup captures pre-swap content
		// (R23a). Materialize then performs the pending/prior rename — by
		// the time .claude/ is overwritten, the snapshot is already on disk.
		// SnapshotForDiscard returns nil when .claude/ doesn't exist (NoActive
		// or active-with-deleted-tree drift): there's nothing to back up, and
		// nil is propagated to the caller as "no snapshot taken" (TS parity).
		backup, err := state.SnapshotForDiscard(opts.Paths)
		if err != nil {
			return ApplyGateResult{}, err
		}
		r, err := state.Materialize(opts.Paths, opts.Plan, opts.Merged, state.MaterializeOptions{}, backup)
		if err != nil {
			// PR25: surface the backup path even when Materialize fails after
			// snapshot. The user's edits are on disk at `*backup`; D7 needs the
			// path to render "your edits were saved at <path>" alongside the
			// failure. Returning a zero-value result here would orphan the
			// snapshot dir without any caller able to find it. When backup is
			// nil (nothing to back up), pass nil through unchanged.
			return ApplyGateResult{
				Action:         ApplyActionAborted,
				BackupSnapshot: backup,
			}, err
		}
		return ApplyGateResult{
			Action:            ApplyActionMaterialized,
			BackupSnapshot:    backup,
			MaterializeResult: &r,
		}, nil

	case GateChoicePersist:
		if opts.ActiveProfileName == "" {
			// Defense-in-depth: DecideGate returns no-drift-proceed when
			// FingerprintOk is false (which implies activeProfile is null).
			// If we land here, a caller bypassed the decider with a hand-
			// crafted choice. Refuse rather than write into a profile-less
			// directory.
			return ApplyGateResult{}, errors.New("persist gate choice requires an active profile in .state.json — none recorded")
		}
		r, err := state.PersistAndMaterialize(opts.Paths, state.PersistOptions{
			ActiveProfileName: opts.ActiveProfileName,
			NewPlan:           opts.Plan,
			NewMerged:         opts.Merged,
		})
		if err != nil {
			return ApplyGateResult{}, err
		}
		return ApplyGateResult{
			Action:            ApplyActionPersistedAndMaterialized,
			BackupSnapshot:    nil,
			MaterializeResult: &r,
		}, nil

	default:
		return ApplyGateResult{}, errors.New("unreachable: unknown gate choice " + string(choice))
	}
}
